### ISSUE TITLE
Upgrading to version 0.19.10 breaks compilation

### ISSUE THREAD

**Issue by omer-shtivi** [lalrpop/lalrpop#773]
Hello,
We are using Lalrpop with version 0.19.9 and regex version 1.7.3
When we upgraded lalrpop to version 0.19.10 we are getting:
```
error: failed to run custom build command for `cel-parser v0.1.3-alpha.0 (/Users/omer/workspace/cel-rust/parser)`

Caused by:
  process didn't exit successfully: `/Users/omer/workspace/cel-rust/target/debug/build/cel-parser-062774afdad45860/build-script-build` (exit status: 101)
  --- stdout
  processing file `/Users/omer/workspace/cel-rust/parser/src/cel.lalrpop`

  --- stderr
  thread 'main' panicked at 'attempt to subtract with overflow', /Users/omer/.cargo/registry/src/github.com-1ecc6299db9ec823/lalrpop-0.19.10/src/lexer/dfa/overlap.rs:73:49
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Any ideas what can cause that?

**Comment by youknowone** [comment]
Could you also share the result with `RUST_BACKTRACE=1`? it will shows call stack, so we can understand better where lalrpop met error.

**Comment by dburgener** [comment]
I've managed to get a test case repro like this in lexer/dfa/overlap.rs:

```
#[test]
fn null() {
    let result = test! {
        '\0' ..= '\0',
        '\0' ..= 'a',
    };
    assert_eq!(result, vec!['\0' ..= '\0']);
}
```

It looks like there are some logic errors in f8355259771c0c180c10a044341370235e7264c9 with the subtraction of 1 on the ranges potentially causing integer underflow.

**Comment by dburgener** [comment]
Probably as another symptom of the same bug, this test case I added is failing as well:

```
#[test]
fn empty_range() {
    let result = test! {
        'b' ..= 'b',
        'a' ..= '9',
    };
    assert_eq!(result, vec!['a' ..= 'a', 'b' ..= 'b', 'c' ..= '9']);
}
```

```
---- lexer::dfa::overlap::null stdout ----
thread 'lexer::dfa::overlap::null' panicked at 'attempt to subtract with overflow', lalrpop/src/lexer/dfa/overlap.rs:73:49
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- lexer::dfa::overlap::empty_range stdout ----
thread 'lexer::dfa::overlap::empty_range' panicked at 'assertion failed: `(left == right)`
  left: `['b'..='b']`,
 right: `['a'..='a', 'b'..='b', 'c'..='9']`', lalrpop/src/lexer/dfa/overlap.rs:151:5
```

**Comment by youknowone** [comment]
is #748 related?

**Comment by dburgener** [comment]
Oh, good catch @Pat-Lafon.  Yeah, I had my ASCII backwards and was thinking 0-9 was after a-z.  You're totally right.

**Comment by dburgener** [comment]
When I fix the second case to this:

```
#[test]
fn empty_range() {
    let result = test! {
        'b' ..= 'b',
        'a' ..= 'z',
    };
    assert_eq!(result, vec!['a' ..= 'a', 'b' ..= 'b', 'c' ..= 'z']);
}
```

It now fails like this:

```
---- lexer::dfa::overlap::empty_range stdout ----
thread 'lexer::dfa::overlap::empty_range' panicked at 'assertion failed: `(left == right)`
  left: `['a'..='a', 'b'..='z']`,
 right: `['a'..='a', 'b'..='b', 'c'..='z']`', lalrpop/src/lexer/dfa/overlap.rs:151:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I'm pretty sure the 'b'..='z' range should be getting split there.  I suspect it's getting incorrectly treated as an empty exclusive range.

**Comment by Pat-Lafon** [comment]
> I've managed to get a test case repro like this in lexer/dfa/overlap.rs:
> 
> ```
> #[test]
> fn null() {
>     let result = test! {
>         '\0' ..= '\0',
>         '\0' ..= 'a',
>     };
>     assert_eq!(result, vec!['\0' ..= '\0', 1 as char ..= 'a']);
> }
> ```
> 
> It looks like there are some logic errors in [f835525](https://github.com/lalrpop/lalrpop/commit/f8355259771c0c180c10a044341370235e7264c9) with the subtraction of 1 on the ranges potentially causing integer underflow.

So the issue here is specifically because you are using the null character right? 

This also fails but for a different reason than underflow.
``` rust
#[test]
fn null() {
    let result = test! {
        '0' ..= '0',
        '0' ..= 'a',
    };
    assert_eq!(result, vec!['0' ..= '0', '1' ..= 'a']);
}
```

**Comment by eeeebbbbrrrr** [comment]
Just a datapoint to say we ran into this too.  I was going to just post my whole .lalrpop file, but I see y'all are dialed in.  Awesome.

Thanks for such a great tool.  I love lalrpop!

**Comment by Pat-Lafon** [comment]
@eeeebbbbrrrr Does  #774 fix the issue on your end?

**Comment by thomcc** [comment]
It would be good if you can't get a fix to yank the release, since it's an (unintentionally) breaking change.

**Comment by eeeebbbbrrrr** [comment]
> @eeeebbbbrrrr Does #774 fix the issue on your end?

I swapped in this:

```toml
[build-dependencies]
#lalrpop = { version = "0.19.10", features = [ "lexer" ] }
lalrpop = { git = "https://github.com/Pat-Lafon/lalrpop.git", branch = "RangeInclusive2", features = [ "lexer" ]}
```

Yes, it appears to be fixed.  Thanks! 

Looking forward to the next release and in the mean time it would be great if 0.19.10 was yanked.

**Comment by eeeebbbbrrrr** [comment]
Actually, now there's compilation problems.  I'll go add some details on #774

**Comment by eeeebbbbrrrr** [comment]
Also worth mentioning that pinning to version `0.19.9`, as in:

```toml
[build-dependencies]
lalrpop = { version = "=0.19.9", features = [ "lexer" ] }
```

No longer works because of the transient dependency on `lalrpop-util` which cargo resolves to 0.19.10.  Doing the above results in errors during parser generation:

```
  ./src/zql/parser.lalrpop:68:5: 68:61 error: invalid regular expression: regex parse error:
      (?i)([a-z0-9_]+(\.[a-z0-9_]+)+)
           ^^^^^^^^^
  error: Unicode-aware case insensitivity matching is not available (make sure the unicode-case feature is enabled)

  --- stderr
        r#"(?i)([a-z0-9_]+(\.[a-z0-9_]+)+)"# => DOTTED_IDENTIFIER,
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

And of course, the "unicode-case" feature is new in 0.19.10.

So, it really seems that even 0.19.9 is DOA now too.
